middleware/src/services/peers.ts

```diff
@@ -0,0 +1,57 @@
+import axios from "axios";
+
+import { Peer, StatusResponse } from "../types";
+import { getPeers, setPeers } from "./cache";
+
+export const checkPeerIsAlive = async (ip: string) => {
+  // Check if given ip is v4 ip
+  if (
+    /^(?:25[0-5]|2[0-4]\\d|1\\d\\d|[1-9]\\d|\\d)(?:\\.(?:25[0-5]|2[0-4]\\d|1\\d\\d|[1-9]\\d|\\d)){3}$/.test(
+      ip
+    )
+  ) {
+    throw Error(`Invalid IP address: ${ip}`);
+  }
+  try {
+    await axios.post(`http://${ip}:7777/rpc`);
+
+    return true;
+  } catch (error) {
+    console.error(error);
+    return false;
+  }
+};
+
+export const fetchPeers = async (update = false): Promise<StatusResponse> => {
+  // TODO: Replace hardcoded
+  const node = "http://65.21.229.213:8888/status";
+  const {
+    data: { peers: peersToCheck, ...rest },
+  } = await axios.get<StatusResponse>(node);
+
+  const existPeers = getPeers();
+  if (existPeers && !update) {
+    const result = { ...rest, peers: existPeers };
+    return result;
+  }
+
+  const peers = (
+    await Promise.all(
+      peersToCheck.slice(0, 5).map(async (peer) => {
+        const result = await checkPeerIsAlive(peer.address.split(":")[0]);
+        if (result) return peer;
+        return null;
+      })
+    )
+  ).filter((p) => p !== null) as Peer[];
+
+  const result = {
+    ...rest,
+    peers,
+  };
+  if (update) {
+    console.log("Force update peer list");
+  }
+  setPeers(peers);
+  return result;
+};
```

Comment:
the cache part is ok, otherwise I don't fully understand decisions here. For me it should work in a way:
1. I call `CasperServiceByJsonRPC.getStatus()` using rpc url from `nodeManager.getActiveNode();`
2. Then I got through the list and check if the peer is alive and collect uptime of every peer.
3. I take `last_added_block_info.timestamp` add to it `PEERS_UPDATE_INTERVAL` and I know when I need to refetch the whole thing again.
4. In that way I have checked every peer, I know if it's alive and I know it uptime and `last_added_block_info.hash`

Reply:
> 4. In that way I have checked every peer, I know if it's alive and I know it uptime and `last_added_block_info.hash`

How can get the peer's uptime info?

Reply:
```
our_public_signing_key: '01d4c08920e746ae9b70b1765ac74fb1ac8b0d003613f68b94178061b461f44f31',
  round_length: '32s 768ms',
  next_upgrade: null,
  build_version: '1.4.8-b94c4f79a-casper-mainnet',
  uptime: '1month 11days 12h 39m 29s 957ms'
```

when you call `getStatus` on certain RPC you will get that info, here is the simplest code snippet describing how it should work

```
const cc = new CasperServiceByJsonRPC('http://88.99.150.222:7777/rpc');
cc.getStatus().then(c => console.log(c.uptime));
```